crates/ra_project_model/src/sysroot.rs

```diff
@@ -93,47 +90,59 @@ impl Sysroot {
+                Some(code) => bail!(
+                    "failed to locate sysroot: rustc --print sysroot exited with code {}",
+                    code
+                ),
+                None => {
+                    bail!("failed to locate sysroot: rustc --print sysroot terminated by signal")
+                }
+            };
+        }
+        let stdout = String::from_utf8(rustc_output.stdout)?;
+        let sysroot_path = Path::new(stdout.trim());
+        Ok(sysroot_path.join("lib/rustlib/src/rust/src"))
     }
 
-    let rustc_output = Command::new("rustc")
-        .current_dir(cargo_toml.parent().unwrap())
-        .args(&["--print", "sysroot"])
-        .output()
-        .context("rustc --print sysroot failed")?;
-    if !rustc_output.status.success() {
-        match rustc_output.status.code() {
-            Some(code) => {
-                bail!("failed to locate sysroot: rustc --print sysroot exited with code {}", code)
-            }
-            None => bail!("failed to locate sysroot: rustc --print sysroot terminated by signal"),
-        };
+    fn try_install_rust_src(cargo_toml: &Path) -> Result<PathBuf> {
+        let rustup_output = Command::new("rustup")
+            .current_dir(cargo_toml.parent().unwrap())
+            .args(&["component", "add", "rust-src"])
+            .output()
+            .context("rustup component add rust-src failed")?;
+        if !rustup_output.status.success() {
+            match rustup_output.status.code() {
+                Some(code) => bail!(
+                    "failed to install rust-src: rustup component add rust-src exited with code {}",
+                    code
+                ),
+                None => bail!(
+                    "failed to install rust-src: rustup component add rust-src terminated by signal"
+                ),
+            };
+        }
+        try_find_src_path(cargo_toml)
     }
-    let stdout = String::from_utf8(rustc_output.stdout)?;
-    let sysroot_path = Path::new(stdout.trim());
-    Ok(sysroot_path.join("lib/rustlib/src/rust/src"))
-}
 
-fn try_install_rust_src(cargo_toml: &Path) -> Result<PathBuf> {
-    let rustup_output = Command::new("rustup")
-        .current_dir(cargo_toml.parent().unwrap())
-        .args(&["component", "add", "rust-src"])
-        .output()
-        .context("rustup component add rust-src failed")?;
-    if !rustup_output.status.success() {
-        match rustup_output.status.code() {
-            Some(code) => bail!(
-                "failed to install rust-src: rustup component add rust-src exited with code {}",
-                code
-            ),
-            None => bail!(
-                "failed to install rust-src: rustup component add rust-src terminated by signal"
-            ),
-        };
+    let src = try_find_src_path(cargo_toml)?;
+    if !src.exists() {
+        try_install_rust_src(cargo_toml)
+    } else {
+        Ok(src)
     }
-    try_find_src_path(cargo_toml)
 }
```

Comment:
Hm, now that this code is in one place, it looks like we can simplify it further?

```
fn get_or_install(cargo_toml) {
    if check_env() { return; }  // this can't be changed afterwards
    
    let sysroot_path = { ..}; // this also we can compute only once

    let src_path = sysroot_path.join("lib/rustlib/src/rust/src");
    if !src_path.exists() { install... }
    if !src_path.exists() { bail!() }
    return
}

```

Reply:
> // this also we can compute only once

This I've failed to realize for some reason, thanks for the pointer.

Reply:
Let me try to extract the common `Command`-related logic also, since we've started it.